src/java.base/share/classes/java/lang/IO.java

```diff
@@ -0,0 +1,196 @@
+        System.out.println();
+    }
+
+    /**
+     * Writes a string representation of the specified object to the
+     * standard output.
+     *
+     * <p> The effect is as if {@link java.io.PrintStream#print(Object) print(obj)}
+     * had been called on {@code System.out}.
+     *
+     * @param obj the object to print, may be {@code null}
+     */
+    public static void print(Object obj) {
+        var out = System.out;
+        out.print(obj);
+        out.flush();
+    }
+
+    /**
+     * Reads a single line of text from the standard input.
+     * <p>
+     * One line is read from the decoded input as if by
+     * {@link java.io.BufferedReader#readLine() BufferedReader.readLine()}
+     * and then the result is returned.
+     * <p>
+     * If necessary, this method first sets up charset decoding, as described in
+     * above in the class specification.
+     *
+     * @return a string containing the line read from the standard input, not
+     * including any line separator characters. Returns {@code null} if an
+     * end of stream has been reached without having read any characters.
+     *
+     * @throws IOError if an I/O error occurs
+     */
+    public static String readln() {
+        try {
+            return reader().readLine();
+        } catch (IOException ioe) {
+            throw new IOError(ioe);
+        }
+    }
+
+    /**
+     * Writes a prompt and then reads a line of input.
+     * <p>
+     * Writes a prompt as if by calling {@code print}, and then reads a single
+     * line of text as if by calling {@link readln readln()}.
+     * <p>
+     * If necessary, this method first sets up charset decoding, as described in
+     * above in the class specification.
+     *
+     * @param prompt the prompt string, may be {@code null}
+     *
+     * @return a string containing the line read from the standard input, not
+     * including any line separator characters. Returns {@code null} if an
+     * end of stream has been reached without having read any characters.
+     *
+     * @throws IOError if an I/O error occurs
+     */
+    public static String readln(String prompt) {
+        print(prompt);
+        return readln();
+    }
+
+    /**
+     * The BufferedReader used by readln(). Initialized under a class lock by
+     * the reader() method.
+     */
+    static BufferedReader br;
+
+    /**
+     * On the first call, creates an InputStreamReader to decode characters from
+     * System.in, wraps it in a BufferedReader, and returns the BufferedReader.
+     * These objects are cached and returned by subsequent calls.
+     *
+     * @return the internal BufferedReader instance
+     */
+    static synchronized BufferedReader reader() {
+        if (br == null) {
+            String enc = System.getProperty("stdin.encoding", "");
+            Charset cs = Charset.forName(enc, StandardCharsets.UTF_8);
+            br = new BufferedReader(new InputStreamReader(System.in, cs));
```

Comment:
I suspect we need to fall back to `stdout.encoding` first before UTF-8 especially in Windows if Java does not set it automatically when launched.

Reply:
I believe that after:
https://github.com/openjdk/jdk/pull/24738
the `stdin.encoding` property is set when the process is connected to a console, which is probably the main case where the encoding can be detected automatically.

Reply:
I misunderstood that I have tested `stdin.encoding`.
I cannot confirm that `stdin.encoding` is changed in Windows to proper ones because it has not been included in the latest EA build (20) and I cannot build master for Windows. However, it is changed to `US-ASCII` in Ubuntu when I change `LANG`. I will believe its author until the latest EA build is updated.

Reply:
Anyway you should merge master first to test `stdin.encoding`.

Reply:
`stdin.encoding` seems to work as intended in Windows for Japanese. No need to change. (I wish the constructors of `InputStreamReader` and `Scanner` used `stdin.encoding` instead of UTF-8 for `System.in` by defaultI) Sorry for having bothered you.

Reply:
Hello @tats-u , there is no bother, I appreciate you looking at this.

The addition of `stdin.encoding` in [JDK-8350703](https://bugs.openjdk.org/browse/JDK-8350703) (#24738) happened very quickly. I worked with @naotoj and @AlanBateman on it and we think it does the right thing, but there are many Windows configurations that we don't have the opportunity to test, so more feedback is helpful.

Initially it would be good to understand if `stdin.encoding` is specified and implemented correctly, and whether it provides the right behavior for `java.lang.IO` in JEP 512.

In the future we can consider additional uses of `stdin.encoding` in other parts of the JDK.

Reply:
`InputStreamReader` and `Scanner` taking `stdin.encoding` is an interesting idea, but would introduce compatiblity concerns. It would have been simpler if we had `stdin/out/err.encoding` system properties prior to JEP 400. 
As to testing on Windows, I would like to know how the Windows `UILanguage` affects these, afaik we only have tested on English Windows. Say how it works on Japanese Windows with English system locale. However, since we have not changed any existing behavior (from the code wise), I don't think we would see any issues.

Reply:
In Java 17 or prior, non-advanced Windows users especially whose languages have common encodings for ANSI/OEM code pages (e.g. Japanese) satisfied the behavior at that time. 
JEP 400 broke everything in stdin in Windows. Currently we need to check both stdin.encoding and stdout.encoding to support older Java versions. 

One idea to fix is to check whether System.in or not in the constructor of InputStreamReader.

https://github.com/tats-u/jdk/commit/3cbc7be971234ed3c52899ef2b16c040dc120418

Reply:
I think we have only to change console encodings by e.g. `chcp 932` (CMD) or `[Console]::InputEncoding = [Console]::OutputEncoding = [System.Text.Encoding]::Get Encoding(932)` (PowerShell) in Windows.